File: src/openjarvis/security/taint.py
"""Taint tracking — information flow control.

Prevents data leakage through tool chains.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, FrozenSet, Optional, Set


class TaintLabel(str, Enum):
    """Labels for tainted data."""
    PII = "pii"
    SECRET = "secret"
    USER_PRIVATE = "user_private"
    EXTERNAL = "external"


@dataclass(frozen=True)
class TaintSet:
    """Immutable set of taint labels attached to data."""
    labels: FrozenSet[TaintLabel] = field(default_factory=frozenset)

    def union(self, other: TaintSet) -> TaintSet:
        """Merge two taint sets."""
        return TaintSet(labels=self.labels | other.labels)

    def has(self, label: TaintLabel) -> bool:
        """Check if a specific label is present."""
        return label in self.labels

    def __bool__(self) -> bool:
        return bool(self.labels)

    @classmethod
    def from_labels(cls, *labels: TaintLabel) -> TaintSet:
        """Create from one or more labels."""
        return cls(labels=frozenset(labels))


# Sink policy: which taint labels are forbidden for each tool
# If a tool appears here, data with any of the listed labels MUST NOT
# be passed to that tool.
SINK_POLICY: Dict[str, Set[TaintLabel]] = {
    "web_search": {TaintLabel.PII, TaintLabel.SECRET},
    "channel_send": {TaintLabel.SECRET},
    "code_interpreter": {TaintLabel.SECRET},
}

# Patterns for auto-detecting taint in text
_PII_PATTERNS = [
    re.compile(r"\b\d{3}-\d{2}-\d{4}\b"),                    # SSN
    re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b"),  # email
    re.compile(r"\b\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4}\b"),  # credit card
    re.compile(r"\b\+?1?\s*\(?[2-9]\d{2}\)?\s*[-.\s]?\d{3}\s*[-.\s]?\d{4}\b"),  # phone
]

_SECRET_PATTERNS = [
    re.compile(r"(?:sk|pk|api)[_-][a-zA-Z0-9]{20,}"),        # API keys
    re.compile(r"(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{36,}"),  # GitHub tokens
    re.compile(r"-----BEGIN (?:RSA |EC |DSA )?PRIVATE KEY-----"),  # Private keys
    re.compile(
        r"(?:bearer|token|password|secret|key)\s*[=:]\s*\S{8,}",
        re.IGNORECASE,
    ),  # Generic secrets
]


def check_taint(tool_name: str, taint: TaintSet) -> Optional[str]:
    """Check if *taint* labels violate the sink policy for *tool_name*.

    Returns a violation description string, or None if clean.
    """
    forbidden = SINK_POLICY.get(tool_name)
    if forbidden is None:
        return None
    violations = taint.labels & forbidden
    if violations:
        labels_str = ", ".join(
            v.value
            for v in sorted(violations, key=lambda x: x.value)
        )
        return (
            f"Data with labels [{labels_str}] "
            f"cannot be sent to '{tool_name}'."
        )
    return None


def declassify(taint: TaintSet, remove: TaintLabel, reason: str) -> TaintSet:
    """Remove a taint label with an explicit reason (for audit).

    The *reason* is not stored on the TaintSet itself but should be
    logged externally for accountability.
    """
    return TaintSet(labels=taint.labels - {remove})


def auto_detect_taint(text: str) -> TaintSet:
    """Auto-detect taint labels in text content.

    Uses regex patterns to detect PII and secrets in tool output.
    """
    labels: set[TaintLabel] = set()

    for pattern in _PII_PATTERNS:
        if pattern.search(text):
            labels.add(TaintLabel.PII)
            break

    for pattern in _SECRET_PATTERNS:
        if pattern.search(text):
            labels.add(TaintLabel.SECRET)
            break

    return TaintSet(labels=frozenset(labels))


def propagate_taint(
    input_taint: TaintSet,
    output_text: str,
) -> TaintSet:
    """Propagate taint: union of input taint with auto-detected output taint."""
    output_taint = auto_detect_taint(output_text)
    return input_taint.union(output_taint)


__all__ = [
    "SINK_POLICY",
    "TaintLabel",
    "TaintSet",
    "auto_detect_taint",
    "check_taint",
    "declassify",
    "propagate_taint",
]

File: repos/openjarvis/rust/crates/openjarvis-security/src/taint.rs
//! Taint tracking — information flow control.

use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
pub enum TaintLabel {
    Pii,
    Secret,
    UserPrivate,
    External,
}

impl std::fmt::Display for TaintLabel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TaintLabel::Pii => write!(f, "pii"),
            TaintLabel::Secret => write!(f, "secret"),
            TaintLabel::UserPrivate => write!(f, "user_private"),
            TaintLabel::External => write!(f, "external"),
        }
    }
}

/// Immutable set of taint labels attached to data.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TaintSet {
    labels: HashSet<TaintLabel>,
}

impl TaintSet {
    pub fn new() -> Self {
        Self {
            labels: HashSet::new(),
        }
    }

    pub fn from_labels(labels: &[TaintLabel]) -> Self {
        Self {
            labels: labels.iter().copied().collect(),
        }
    }

    pub fn union(&self, other: &TaintSet) -> TaintSet {
        TaintSet {
            labels: self.labels.union(&other.labels).copied().collect(),
        }
    }

    pub fn has(&self, label: TaintLabel) -> bool {
        self.labels.contains(&label)
    }

    pub fn is_empty(&self) -> bool {
        self.labels.is_empty()
    }

    pub fn labels(&self) -> &HashSet<TaintLabel> {
        &self.labels
    }

    pub fn remove(&self, label: TaintLabel) -> TaintSet {
        let mut new_labels = self.labels.clone();
        new_labels.remove(&label);
        TaintSet { labels: new_labels }
    }
}

/// Sink policy: which taint labels are forbidden for each tool.
static SINK_POLICY: Lazy<HashMap<&'static str, HashSet<TaintLabel>>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert(
        "web_search",
        HashSet::from([TaintLabel::Pii, TaintLabel::Secret]),
    );
    m.insert("channel_send", HashSet::from([TaintLabel::Secret]));
    m.insert("code_interpreter", HashSet::from([TaintLabel::Secret]));
    m
});

static PII_PATTERNS: Lazy<Vec<Regex>> = Lazy::new(|| {
    vec![
        Regex::new(r"\b\d{3}-\d{2}-\d{4}\b").unwrap(),
        Regex::new(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b").unwrap(),
        Regex::new(r"\b\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4}\b").unwrap(),
        Regex::new(r"\b\+?1?\s*\(?[2-9]\d{2}\)?\s*[-.\s]?\d{3}\s*[-.\s]?\d{4}\b").unwrap(),
    ]
});

static SECRET_PATTERNS: Lazy<Vec<Regex>> = Lazy::new(|| {
    vec![
        Regex::new(r"(?:sk|pk|api)[_-][a-zA-Z0-9]{20,}").unwrap(),
        Regex::new(r"(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{36,}").unwrap(),
        Regex::new(r"-----BEGIN (?:RSA |EC |DSA )?PRIVATE KEY-----").unwrap(),
        Regex::new(r"(?i)(?:bearer|token|password|secret|key)\s*[=:]\s*\S{8,}").unwrap(),
    ]
});

/// Check if taint labels violate the sink policy for a tool.
/// Returns a violation description, or None if clean.
pub fn check_taint(tool_name: &str, taint: &TaintSet) -> Option<String> {
    let forbidden = SINK_POLICY.get(tool_name)?;
    let violations: Vec<_> = taint
        .labels
        .intersection(forbidden)
        .collect();
    if violations.is_empty() {
        return None;
    }
    let mut sorted: Vec<_> = violations.iter().map(|v| v.to_string()).collect();
    sorted.sort();
    Some(format!(
        "Data with labels [{}] cannot be sent to '{}'.",
        sorted.join(", "),
        tool_name
    ))
}

/// Auto-detect taint labels in text content.
pub fn auto_detect_taint(text: &str) -> TaintSet {
    let mut labels = HashSet::new();

    for pat in PII_PATTERNS.iter() {
        if pat.is_match(text) {
            labels.insert(TaintLabel::Pii);
            break;
        }
    }

    for pat in SECRET_PATTERNS.iter() {
        if pat.is_match(text) {
            labels.insert(TaintLabel::Secret);
            break;
        }
    }

    TaintSet { labels }
}

/// Propagate taint: union of input taint with auto-detected output taint.
pub fn propagate_taint(input_taint: &TaintSet, output_text: &str) -> TaintSet {
    let output_taint = auto_detect_taint(output_text);
    input_taint.union(&output_taint)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_taint_set_operations() {
        let t1 = TaintSet::from_labels(&[TaintLabel::Pii]);
        let t2 = TaintSet::from_labels(&[TaintLabel::Secret]);
        let merged = t1.union(&t2);
        assert!(merged.has(TaintLabel::Pii));
        assert!(merged.has(TaintLabel::Secret));
    }

    #[test]
    fn test_check_taint_violation() {
        let taint = TaintSet::from_labels(&[TaintLabel::Pii, TaintLabel::Secret]);
        let result = check_taint("web_search", &taint);
        assert!(result.is_some());
    }

    #[test]
    fn test_check_taint_clean() {
        let taint = TaintSet::from_labels(&[TaintLabel::External]);
        let result = check_taint("web_search", &taint);
        assert!(result.is_none());
    }

    #[test]
    fn test_auto_detect_pii() {
        let taint = auto_detect_taint("SSN: 123-45-6789");
        assert!(taint.has(TaintLabel::Pii));
    }

    #[test]
    fn test_auto_detect_secret() {
        let taint = auto_detect_taint("key: sk-abcdefghijklmnopqrstuvwxyz");
        assert!(taint.has(TaintLabel::Secret));
    }

    #[test]
    fn test_propagate() {
        let input = TaintSet::from_labels(&[TaintLabel::External]);
        let result = propagate_taint(&input, "SSN: 123-45-6789");
        assert!(result.has(TaintLabel::External));
        assert!(result.has(TaintLabel::Pii));
    }
}
